File: javascript/palindrome-products/palindrome-products.js
export class Palindromes {
  static generate(factors) {
    if (factors.minFactor > factors.maxFactor) {
      throw new Error('min must be <= max');
    };
    let palindromesObject = {};
    for (let i = factors.minFactor; i <= factors.maxFactor; i++) {
      for (let j = i; j <= factors.maxFactor; j++) {
        if (isPalindrome(i * j)) {
          if (!palindromesObject[i * j]) {
            palindromesObject[i * j] = [[i, j]];
          } else {
            palindromesObject[i * j] = palindromesObject[i * j].concat([[i, j]]);
          };
        };
      };
    };
    let palindromesArray = [];
    for (let palindrome in palindromesObject) {
      palindromesArray.push({value: Number(palindrome), factors: palindromesObject[palindrome]});
    };
    palindromesArray.sort((p1, p2) => {
      return p1.value - p2.value;
    });
    let extremes = {smallest: {value: null, factors: []}, largest: {value: null, factors: []}};
    if (palindromesArray.length > 0) {
      extremes.smallest = palindromesArray[0];
      extremes.largest = palindromesArray[palindromesArray.length - 1];
    };
    return extremes;
  }
}

function isPalindrome(number) {
  let stringNumber = number.toString()
  for (let i = 0; i < stringNumber.length / 2; i++) {
    if (stringNumber[i] !== stringNumber[stringNumber.length - i - 1]) {
      return false;
    };
  };
  return true;
}

File: python/palindrome-products/palindrome_products.py
def largest(max_factor, min_factor = 0):
    """Given a range of numbers, find the largest palindromes which
       are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
             Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if min_factor > max_factor:
        raise ValueError("min must be <= max")
    value, factors = -1, []
    for i in range(max_factor, min_factor - 1, -1):
        if i ** 2 < value:
            break
        for j in range(i, min_factor - 1, -1):
            digits = [digit for digit in str(i * j)]
            if digits == digits[::-1]:
                if value == i * j:
                    factors.append([i, j])
                elif value < i * j:
                    value = i * j
                    factors = [[i, j]]
    return ((value if value != -1 else None), factors)
            


def smallest(max_factor, min_factor = 0):
    """Given a range of numbers, find the smallest palindromes which
    are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
    Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if min_factor > max_factor:
        raise ValueError("min must be <= max")
    value, factors = max_factor ** 2 + 1, []
    for i in range(min_factor, max_factor + 1):
        if i ** 2 > value:
            break
        for j in range(i, max_factor + 1):
            digits = [digit for digit in str(i * j)]
            if digits == digits[::-1]:
                if value == i * j:
                    factors.append([i, j])
                elif value > i * j:
                    value = i * j
                    factors = [[i, j]]
    return ((value if value != max_factor ** 2 + 1 else None), factors)
